### ISSUE TITLE
Opcode 0xC4 (CALL NZ,a16) doesn't push current PC value + 3 to the stack

### ISSUE THREAD

**Issue by dylmeadows** [darrena092/DMGe#40]
As the title says, opcode 0xC4 is coded to behave as a JUMP when it should be pushing the current PC value + 3 to the stack before setting PC to a16.

**Comment by Quaker762** [comment]
Wew, you're totally right. I wrote this when I was new to C/C++ and had no idea what I was doing. I got confused and set flags willy nilly/performed weird jumps. Definitely keen to work on a completely fresh version if you and @darrena092 are as well.

**Comment by dylmeadows** [comment]
I'm currently in my senior year at university, but I would love to help wherever I can. I've been working on my own GameBoy emulator written in Java, but had to slow down some of my progress to focus on school. I'm not as fluent in C++ as I am in Java, but I do have experience with it.

**Comment by darrena092** [comment]
Sorry I'm not chiming in on this much - pretty busy at the minute!

I'm not particularly interested on reviving this project TBH, as I don't think there's much of a need for another Gameboy emulator, and I don't think I'd personally learn much by repeating this exercise.

I would however be interested in writing a new emulator that targets a different platform. Perhaps the playstation 1 - something with 3d rendering, and also a MIPS core, as that's what I've been studying in uni recently.

Either of you up for that? We could also work in Java, if you'd like.

Darren

**Comment by dylmeadows** [comment]
Playstation would be cool. I'd be down for that or an N64 emulator

**Comment by Quaker762** [comment]
[How about the original XBOX??](http://www.neogaf.com/forum/showpost.php?p=48088464&postcount=26)? Nah, just kidding haha.

I'd definitely be down to do a PS1 emulator. It would probably be best to do it in C or C++ so we can make a dynarec and [emit our own assembly](https://github.com/marco9999/Dynarec_Guide/blob/master/Introduction%20to%20Dynamic%20Recompilation%20in%20Emulation.pdf). I'm down for using Java, but I'm not too sure if we'll experience any significant slowdowns as we'd probably have to do a fetch-decode-execute loop, which is too slow for a system like this (though with modern CPUs I'm not too sure if this is the case). That and I really dislike the lack of struct and direct memory copying. For example, [this program](https://github.com/Palm-Studios/sh3tools/blob/master/source/000vid/000vid.c) I ported from Java was sped up by ~700% due to the ability to copy entire blocks of memory. I could be wrong though, I haven't used Java in a loooooong time.

> I'm not particularly interested on reviving this project TBH, as I don't think there's much of a need for another Gameboy emulator, and I don't think I'd personally learn much by repeating this exercise.

I totally agree, though we didn't end up finishing this I think we understand enough about the GameBoy to last a life time [laughs].

> I'd be down for that or an N64 emulator

This would be cool as well, I know there is a load of documentation on the chips because an SGI employee leaked all the documentation in the earl 00's.

If anyone wants to start an organisation/repo, go right ahead, I've got another month before I go back to uni haha.

**Comment by dylmeadows** [comment]
Oops

**Comment by Quaker762** [comment]
> Oops

Lol, I've done this so many times

@darrena092 @dylmeadows Did you actually want to get started on a new emulator when everyone has time to discuss design etc?

**Comment by dylmeadows** [comment]
Yeah I'm game. Is there a better way to communicate with everyone besides the issues selection of the repository?

**Comment by Quaker762** [comment]
IRC, Discord etc?

**Comment by darrena092** [comment]
I'll get a slack group set up and send you guys an invite.

**Comment by Quaker762** [comment]
@darrena092 Awesome can't wait to get started.
